### DISCUSSION TITLE
Activate PHP on macOS Monterrey

### DISCUSSION THREAD

**Discussion by avivamientodev** [shivammathur/homebrew-php#491] [Q&A / Help] [answered]
macOS Monterrey ships without PHP (it's been deprecated). I installed PHP from here using brew, but now I can't find how to activate PHP, it's not being interpreted, all I get is the code from PHP documents. I will appreciate some help. Thanks !!

**Comment by shivammathur** [comment, marked as answer]
After installing php make sure you have linked the PHP version as documented in the README.
https://github.com/shivammathur/homebrew-php#install-php

Then check that the correct PHP version is installed by running `php -v` in a new terminal.

If you are trying to run PHP with a web server like `apache/httpd`. Make sure you have set it up correctly. 
This is a good guide for that: https://getgrav.org/blog/macos-bigsur-apache-multiple-php-versions

**Reply by avivamientodev** [reply]
Thank you very much

**Comment by avivamientodev** [comment]
What I found (it's my recommendation for every one), Is to  go through this process after a clean macOS Big Sur installation, This one is the second time that I use this procedure, and in both cases, installing over a clean installation (before installing anything else), it worked perfectly ... What I would like to know now is if someone after finishing this process,  has installed macOS Monterrey without losing the web server and/or PHP.

**Reply by gemkev** [reply]
I had PHP working perfectly on Big Sur yesterday. Today, I upgraded to Monterey, and it no longer works. I used Homebrew to uninstall/reinstall just in case, and followed ALL of the steps outlined elsewhere. It still doesn't work.

If I type **sudo apachectl -k restart** in the Terminal, I see this message: 

[Tue Oct 26 14:00:32.059085 2021] [so:error] [pid 41267] AH06665: No code signing authority for module at /opt/homebrew/opt/php/lib/httpd/modules/libphp.so specified in LoadModule directive.
httpd: Syntax error on line 194 of /private/etc/apache2/httpd.conf: Code signing absent - not loading module at: /opt/homebrew/opt/php/lib/httpd/modules/libphp.so

I subsequently read that code signing was optional/recommended on Big Sur, but obligatory on Monterey. 

Is there a way around this problem? I'm completely stuck. :) 

Thanks, 
Kevin

**Reply by shivammathur** [reply]
@gemkev

I currently do not have a CI machine for Monterey builds. I will try to set up one.
Till then, please try reinstalling PHP along with its dependencies, and let me know if that fixes it

```zsh
brew update
brew reinstall $(brew deps shivammathur/php/php) shivammathur/php/php
```

**Reply by shivammathur** [reply]
@gemkev 

If the above does not work, try creating a self-signed certificate and using that.

```zsh
codesign --sign "php key" --force --keychain ~/Library/Keychains/login.keychain-db /opt/homebrew/opt/php/lib/httpd/modules/libphp.so
```

Then in `httpd.conf` add the key to the LoadModule line.

```conf
LoadModule php_module /opt/homebrew/opt/php/lib/httpd/modules/libphp.so "php key"
```

Finally, restart Apache.

```zsh
brew services restart httpd
```

Let me know if this works.

**Reply by gemkev** [reply]
Thanks for the quick reply, Shivam. 

I'd already created a self-signed certificate (thanks to another site, which suggested that) and run codesign. But I **hadn't** put the 'php key' at the end of LoadModule line in httpd.conf. 

Once I did, **it worked!**

Just to check whether the first solution would have done the job as well, I ran update/reinstall, and commented out the 'php key' from LoadModule, and then restarted. That **didn't** work. 

So it's a code signing thing. I've now reverted to solution 1, so all good. But I'm wondering what happens when I upgrade to a new version of PHP - do I have to repeat the process?

---
Unrelated question (and betraying my ignorance here!) - what's the difference between restarting the server using **sudo apachectl restart**  and **brew services restart httpd**? 

---

Thanks,  
Kevin

**Reply by shivammathur** [reply]
I'm not sure about reinstallation currently, It's early days for this, so let see. Maybe there is something homebrew can do as part of the installation to avoid this. Otherwise, I will update the README accordingly once I have tested this on Monterey.

`brew services` for `httpd` is a wrapper around `apachectl`.

**Reply by dlakmalb** [reply]
> Thanks for the quick reply, Shivam.
> 
> I'd already created a self-signed certificate (thanks to another site, which suggested that) and run codesign. But I **hadn't** put the 'php key' at the end of LoadModule line in httpd.conf.
> 
> Once I did, **it worked!**
> 
> Just to check whether the first solution would have done the job as well, I ran update/reinstall, and commented out the 'php key' from LoadModule, and then restarted. That **didn't** work.
> 
> So it's a code signing thing. I've now reverted to solution 1, so all good. But I'm wondering what happens when I upgrade to a new version of PHP - do I have to repeat the process?
> 
> Unrelated question (and betraying my ignorance here!) - what's the difference between restarting the server using **sudo apachectl restart** and **brew services restart httpd**?
> 
> Thanks, Kevin

@gemkev can i have the link you used.

**Reply by dlakmalb** [reply]
> @gemkev
> 
> If the above does not work, try creating a self-signed certificate and using that.
> 
> ```shell
> codesign --sign "php key" --force --keychain ~/Library/Keychains/login.keychain-db /opt/homebrew/opt/php/lib/httpd/modules/libphp.so
> ```
> 
> Then in `httpd.conf` add the key to the LoadModule line.
> 
> ```
> LoadModule php_module /opt/homebrew/opt/php/lib/httpd/modules/libphp.so "php key"
> ```
> 
> Finally, restart Apache.
> 
> ```shell
> brew services restart httpd
> ```
> 
> Let me know if this works.

@shivammathur  what should i provide for the php key.

**Reply by gemkev** [reply]
> @gemkev can i have the link you used.

Hey @dlakmalb. Here's the link: [https://www.simplified.guide/macos/apache-php-homebrew-codesign](url).

It takes you through the process step by step, linking to other pages for creating a Certificate Authority and creating a certificate.

I had absolutely no clue how to do any of this, but the rundown at that link is really clear and precise. Follow every step to the letter. Do make sure to scroll step 8 to the right, so you can see how to include the certificate name ('php key' in @shivammathur's answer above). I missed that first time round, as I said in my original reply. 

It's quite involved, but it works perfectly. :)

**Reply by shivammathur** [reply]
@dlakmalb @gemkev "php key" can be any string. It is the certificate name.

**Reply by dlakmalb** [reply]
> I'm not sure about reinstallation currently, It's early days for this, so let see. Maybe there is something homebrew can do as part of the installation to avoid this. Otherwise, I will update the README accordingly once I have tested this on Monterey.
> 
> `brew services` for `httpd` is a wrapper around `apachectl`.

@shivammathur  Do i need to create certificate again after changing the php version 7.4 to 7.1. please help.

**Reply by shivammathur** [reply]
Yes, you will have to run the codesign command again for the PHP 7.1 library and replace it in the apache config with the certificate name as a suffix. You can use the same certificate to sign the 7.1 library.

```zsh
codesign --sign "php key" --force --keychain ~/Library/Keychains/login.keychain-db /opt/homebrew/opt/php@7.1/lib/httpd/modules/libphp7.so
```

**Comment by dlakmalb** [comment]
@shivammathur and @gemkev Thanks for your opinions. I did it. Thank you very much again.
